shoulder/generator/c_header_generator.py

```diff
@@ -21,463 +21,767 @@
+            regs = transforms["remove_reserved_1"].transform(regs)
+            regs = transforms["remove_reserved_sign_extended"].transform(regs)
+            regs = transforms["remove_implementation_defined"].transform(regs)
+            regs = transforms["special_to_underscore"].transform(regs)
+            regs = transforms["remove_coprocessor_am"].transform(regs)
+            regs = transforms["remove_memory_mapped_am"].transform(regs)
+            regs = transforms["remove_system_vector_am"].transform(regs)
+            regs = transforms["remove_system_banked_am"].transform(regs)
+            regs = transforms["remove_system_immediate_am"].transform(regs)
+            regs = transforms["remove_redundant_fields"].transform(regs)
+            regs = transforms["unique_fieldset_names"].transform(regs)
+
+            regs = filters["no_access_mechanism"].filter_exclusive(regs)
+
             outfile_path = os.path.abspath(os.path.join(outpath, "shoulder.h"))
             logger.info("Generating C Header: " + str(outfile_path))
             with open(outfile_path, "w") as outfile:
-                self._generate(outfile, objects)
+                self._generate(outfile, regs)
 
         except Exception as e:
             msg = "{g} failed to generate output {out}: {exception}".format(
-                g = str(type(self).__name__),
-                out = outpath,
-                exception = e
-            )
+                g=str(type(self).__name__),
+                out=outpath,
+                exception=e)
             raise ShoulderGeneratorException(msg)
 
-    @license
-    @include_guard
-    @header_depends
-    def _generate(self, outfile, objects):
-        for obj in objects:
-            if(isinstance(obj, Register)):
-                if not obj.is_sysreg: return
-
-                logger.debug("Writing register: " + str(obj.name))
-                self._generate_register(outfile, obj)
-            else:
-                msg = "Cannot generate object of unsupported {t} type".format(
-                    t = type(obj)
-                )
-                raise ShoulderGeneratorException(msg)
-
-    def _generate_register(self, outfile, reg):
-        self._generate_register_comment(outfile, reg)
-        self._generate_sysreg_get(outfile, reg)
-        if reg.is_writable:
-            self._generate_sysreg_set(outfile, reg)
-
-        outfile.write("\n")
-        self._generate_register_fieldsets(outfile, reg)
-
-    def _generate_register_fieldsets(self, outfile, reg):
-        if not reg.is_sysreg: return
-
-        fieldsets = reg.fieldsets
-        if len(fieldsets) == 1:
-            self._generate_fieldset(outfile, reg, fieldsets[0])
-        elif len(fieldsets) > 1:
-            for idx, fieldset in enumerate(fieldsets):
+    @shoulder.gadget.license
+    @shoulder.gadget.include_guard
+    @shoulder.gadget.header_depends
+    def _generate(self, outfile, regs):
```

Comment:
I'm really a fan of having the register, field, and fieldset generators combined into one function like this